### PR TITLE
TSDB: unicode dimensions

### PR DESCRIPTION
**Pull request by nik9000**
This fixes dimensions with unicode characters in their name. It also
make the sort order of these dimensions something we can explain:
convert all dimensions to utf-8 and sort by that.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/index/mapper/LuceneDocument.java [hunk 1]
```diff
@@ -36,7 +36,7 @@
      * for generating the _tsid field. The map will be used by {@link TimeSeriesIdFieldMapper}
      * to build the _tsid field for the document.
      */
-    private SortedMap<String, BytesReference> dimensionBytes;
+    private SortedMap<BytesRef, BytesReference> dimensionBytes;
```
Comment (nik9000) [inline server/src/main/java/org/elasticsearch/index/mapper/LuceneDocument.java:39]: I switched to bytes ref here to make it really obvious these are sorted by utf-8.

Comment (nik9000) [reply server/src/main/java/org/elasticsearch/index/mapper/LuceneDocument.java:39]: Ulterior motive: I'll want them in utf-8 in more than one place.

File: server/src/main/java/org/elasticsearch/index/mapper/TimeSeriesIdFieldMapper.java [hunk 2]
```diff
@@ -166,19 +167,22 @@ protected String contentType() {
...
+            BytesRef fieldName = entry.getKey();
+            if (fieldName.length > DIMENSION_NAME_LIMIT) {
                 throw new IllegalArgumentException(
-                    "Dimension name must be less than [" + DIMENSION_NAME_LIMIT + "] bytes but [" + fieldName + "] was [" + len + "]."
+                    String.format(
+                        Locale.ROOT,
+                        "Dimension name must be less than [%s] bytes but [%s] was [%s].",
+                        DIMENSION_NAME_LIMIT,
+                        fieldName.utf8ToString(),
+                        fieldName.length
```
Comment (nik9000) [inline server/src/main/java/org/elasticsearch/index/mapper/TimeSeriesIdFieldMapper.java:181]: The formatter made a mess of this when I made the parameters a little longer so I switched it to String.format so human can read the message.

File: server/src/main/java/org/elasticsearch/index/mapper/TimeSeriesIdFieldMapper.java [hunk 3]
```diff
@@ -193,7 +197,7 @@ public static void encodeTsid(StreamOutput out, SortedMap<String, BytesReference
             Map<String, Object> result = new LinkedHashMap<String, Object>(size);
 
             for (int i = 0; i < size; i++) {
-                String name = in.readString();
+                String name = in.readBytesRef().utf8ToString();
```
Comment (nik9000) [inline server/src/main/java/org/elasticsearch/index/mapper/TimeSeriesIdFieldMapper.java:200]: Here's the bug that made this not work - sometimes BytesRef and String don't write the same.

File: server/src/main/java/org/elasticsearch/index/mapper/TimeSeriesIdFieldMapper.java [hunk 4]
```diff
@@ -166,19 +167,22 @@ protected String contentType() {
...
-            int len = fieldNameBytes.length;
-            if (len > DIMENSION_NAME_LIMIT) {
+        for (Map.Entry<BytesRef, BytesReference> entry : dimensionFields.entrySet()) {
+            BytesRef fieldName = entry.getKey();
+            if (fieldName.length > DIMENSION_NAME_LIMIT) {
                 throw new IllegalArgumentException(
-                    "Dimension name must be less than [" + DIMENSION_NAME_LIMIT + "] bytes but [" + fieldName + "] was [" + len + "]."
+                    String.format(
+                        Locale.ROOT,
+                        "Dimension name must be less than [%s] bytes but [%s] was [%s].",
```
Comment (csoulios) [inline server/src/main/java/org/elasticsearch/index/mapper/TimeSeriesIdFieldMapper.java:178]: nit: `fieldName.length` should be formatted as `%d`. Not that `%s` won't work, but usually I prefer to set the correct formatter type.

**Review by csoulios — APPROVED** [review]
LGTM!

File: server/src/main/java/org/elasticsearch/index/mapper/TimeSeriesIdFieldMapper.java (hunk 4 above)
Comment (nik9000) [reply server/src/main/java/org/elasticsearch/index/mapper/TimeSeriesIdFieldMapper.java]: :+1:

File: server/src/main/java/org/elasticsearch/index/mapper/TimeSeriesIdFieldMapper.java (hunk 3 above)
Comment (imotov) [reply server/src/main/java/org/elasticsearch/index/mapper/TimeSeriesIdFieldMapper.java:200]: I wonder if there is a value in hiding the fact that it's BytesRef all the way down and can this hurt us beyond serialization thing.

File: server/src/test/java/org/elasticsearch/index/mapper/TimeSeriesIdFieldMapperTests.java [hunk 5]
```diff
@@ -123,6 +125,23 @@ public void testStrings() throws IOException {
         );
     }
 
+    public void testUnicodeKeys() throws IOException {
+        String fire = new String(new int[] { 0x1F525 }, 0, 1);
+        String coffee = "\u2615";
```
Comment (imotov) [inline server/src/test/java/org/elasticsearch/index/mapper/TimeSeriesIdFieldMapperTests.java:130]: ☕  is a good dimension

File: server/src/main/java/org/elasticsearch/index/mapper/TimeSeriesIdFieldMapper.java (hunk 3 above)
Comment (nik9000) [reply server/src/main/java/org/elasticsearch/index/mapper/TimeSeriesIdFieldMapper.java:200]: Are you thinking this should be it's own class? Like one that just wraps `BytesRef` or something?

Comment (imotov) [reply server/src/main/java/org/elasticsearch/index/mapper/TimeSeriesIdFieldMapper.java:200]: Well, if I am completely honest, I was thinking why do we have 2 different maps with String in one case and BytesRef in another case that essentially represent the same data. But I think I like your interpretation much better. It has quite a bit of complexity that could probably benefit from some encapsulation.

Comment (imotov) [reply server/src/main/java/org/elasticsearch/index/mapper/TimeSeriesIdFieldMapper.java:200]: Just to be clear, maybe we need a TimeSeriesId class to represent this sort of data.

Comment (nik9000) [reply server/src/main/java/org/elasticsearch/index/mapper/TimeSeriesIdFieldMapper.java:200]: Would you like me to do it in this or a followup?

Comment (imotov) [reply server/src/main/java/org/elasticsearch/index/mapper/TimeSeriesIdFieldMapper.java:200]: Yeah, let's not mix refactoring with the actual change.

**Comment by weizijun** [comment]
hi, @nik9000 , as the type of lucene field name is `String`, what's the use case for the key of tsid using `unicode characters`, Using the `ByteRef` to a Map key is a bit dangerous，as ByteRef is a reference, `hashCode` can be changed from its content, it may result in some strange behavior. 
Maybe @jpountz can take a look at this PR?

**Comment by nik9000** [comment]
`BytesRef` is always a little more dangerous because you don't really know what folks are giving you. In this case we only need `BytesRef` for the duration of the method that receives it. Since we're not storing any references to the references I'd say it's save. But, even so, I opened #83799 to hide the `BytesRef`s from plain sight.

**Comment by weizijun** [comment]
> as the type of lucene field name is `String`, what's the use case for the key of tsid using `unicode characters`?

hi, @nik9000 can you explain it? I would like to know the reason why `String` does not support storing field names.

**Comment by nik9000** [comment]
> I would like to know the reason why String does not support storing field names.

It does. But it won't necessarily sort them in the same way as encoding the dimensions to utf-8 and then sorting them. I mean, there was a bug, but we don't need to invoke BytesRef in the Map to solve the bug. But I was mostly concerned about the sorting here. Mostly out of paranoia too. I think it'd be fairly hard to come up with field names that don't sort the same way. But we encode the data using utf-8 anyway so sorting by utf-8 was easy.

**Pull request merged**